ethcore/src/client/client.rs

```diff
@@ -1873,6 +1873,12 @@ impl BlockChainClient for Client {
 	}
 
 	fn filter_traces(&self, filter: TraceFilter) -> Option<Vec<LocalizedTrace>> {
+		let tracedb = self.tracedb.read();
```

Comment:
In all methods it changes lock order and may introduce deadlock with `restore_db` (where we lock `chain_db` and `trace_db` for writing)

Might be better to release the locks as fast as possible.